src/iiif/image_handling.py

```diff
@@ -123,3 +126,96 @@ def scale_image(content, source_file, scaling, content_type):
     scaled_image_data = image_stream.getvalue()
 
     return scaled_image_data
+
+# TODO: Support percentage string: pct:x,y,w,h
+def parse_region_string(region):
+    """
+    Parse the region string from the url (either 'full' or 'x,y,w,h' in pixels)
+
+    :param region: The region string from the url
+    :return: Tuple containing the desired x, y, width and height.
+    """
+    try:
+        parts = region.split(",")
+        if len(parts) != 4:
+            raise ValueError("Invalid format for region")
+        
+        if not (parts[0] or parts[1] or parts[2] or parts[3]):
+            raise ValueError('Invalid format for region')
+
+        desired_x = int(parts[0])
+        desired_y = int(parts[1])
+        desired_width = int(parts[2])
+        desired_height = int(parts[3])
+    except ValueError:
+        raise tools.ImmediateHttpResponse(
+            response=HttpResponse(MALFORMED_REGION_PARAMETER, status=400)
+        )
+
+    return desired_x, desired_y, desired_width, desired_height
+
+# TODO: Move to an util file
+def clamp(n, minn, maxn):
+    return max(min(maxn, n), minn)
+
+# TODO: Extract sub functions to own functions for:
+# - asserting the region is valid
+# - return original when no crop was applied
+def crop_image(content, source_file, region, content_type):
+    """
+    Crop the image to the desired size. Never crop outside the image.
+
+    :param content: The image data
+    :param source_file: Bool whether the raw source file should be returned without any cropping
+    :param region: The region string from the url
+    :param content_type: The content type of the image
+    :return: The cropped image data
+    """
+    if source_file or region.lower() == "full":
+        return content
+
+    desired_x, desired_y, desired_width, desired_height = parse_region_string(region)
+    
+    img = Image.open(BytesIO(content))
+
+    region_has_no_width = desired_width <= 0 or desired_width == None
+    region_has_no_height = desired_height <= 0 or desired_height == None
+    if region_has_no_width or region_has_no_height:
+        raise tools.ImmediateHttpResponse(
+            response=HttpResponse(NON_OVERLAPPING_REGION_PARAMETER, status=400)
```

Comment:
Wat bedoel je precies met de error message dat de region niet overlapt? Hij test inderdaad of er uberhaupt wel een pixel breedte of wijdte is, maar wat heeft dat met overlappen te maken?

Reply:
Because the cropped region has no width or height, there is no part of the cropped image that would overlap with the main image (the cropped image is empty so that would be impossible). According to spec, that is not allowed.

> If the requested region’s height or width is zero, or if the region is entirely outside the bounds of the reported dimensions, then the server should return a 400 (Bad Request) status code.